tutorials/colors/colormap-manipulation.py

```diff
@@ -0,0 +1,198 @@
+use a modest value of 12 so there are not a lot of values to look at.
+"""
+
+import numpy as np
+import matplotlib as mpl
+import matplotlib.pyplot as plt
+from matplotlib import cm
+from matplotlib.colors import ListedColormap, LinearSegmentedColormap
+from collections import OrderedDict
+
+viridis = cm.get_cmap('viridis', 12)
+print(viridis)
+
+##############################################################################
+# The object ``viridis`` is a callable, that when passed a float between
+# 0 and 1 returns an RGBA value from the colormap:
+
+print(viridis(0.56))
+
+##############################################################################
+# The list of colors that comprise the colormap can be directly accessed using
+# the ``colors`` property,
+# or it can be acccessed indirectly by calling  ``viridis`` with an array
+# of values matching the length of the colormap.  Note that the returned list
+# is in the form of an RGBA Nx4 array, where N is the length of the colormap.
+
+print('viridis.colors', viridis.colors)
+print('viridis(range(12))', viridis(range(12)))
+print('viridis(np.linspace(0, 1, 12))', viridis(np.linspace(0, 1, 12)))
+
+##############################################################################
+# The colormap is a lookup table, so "oversampling" the colormap returns
+# nearest-neighbor interpolation (note the repeated colors in the list below)
+
+print('viridis(np.linspace(0, 1, 15))', viridis(np.linspace(0, 1, 15)))
+
+##############################################################################
+# Creating a new ListedColormap
+# =============================
+#
+# This is essential the inverse operation of the above where we supply a
+# Nx4 numpy array with all values between 0 and 1,
+# to `.ListedColormap` to make a new colormap.  This means that
+# any numpy operations that we can do on a Nx4 array make carpentry of
+# new colormaps from existing colormaps quite straight forward.
+#
+# Suppose we want to make the first 25 entries of a 256-length "viridis"
+# colormap pink for some reason:
+
+viridis = cm.get_cmap('viridis', 256)
+newcolors = viridis(np.linspace(0, 1, 256))
+pink = np.array([248/256, 24/256, 148/256, 1])
+newcolors[:25, :] = pink
+newcmp = ListedColormap(newcolors)
+
+
+def plot_examples(cms):
+    """
+    helper function to plot two colormaps
+    """
+    np.random.seed(19680801)
+    data = np.random.randn(30, 30)
+
+    fig, axs = plt.subplots(1, 2, figsize=(6, 3), constrained_layout=True)
+    for [ax, cmap] in zip(axs, cms):
+        psm = ax.pcolormesh(data, cmap=cmap, rasterized=True, vmin=-4, vmax=4)
+        fig.colorbar(psm, ax=ax)
+    plt.show()
+
+plot_examples([viridis, newcmp])
+
+##############################################################################
+# We can easily reduce the range of a colormap; here we choose the middle
+# 0.5 of the colormap.
+
+viridis = cm.get_cmap('viridis', 256)
+newcmp = ListedColormap(viridis(np.linspace(0.25, 0.75, 256)))
```

Comment:
Will this have over-sampling issues?

Reply:
Ooops, yes it will.  Not really noticeable at 256 resolution, but would certainly be a problem for small colormaps.  Changed to 

```python
# We can easily reduce the dynamic range of a colormap; here we choose the middle
# 0.5 of the colormap.  However, we need to interpolate from a larger
# colormap, otherwise the new colormap will have repeated values.

viridisBig = cm.get_cmap('viridis', 512)
newcmp = ListedColormap(viridisBig(np.linspace(0.25, 0.75, 256)))
plot_examples([viridis, newcmp])
```